Add `Simple` module

yawaramin opened the pull request:
Add a module that introduces simplified registration & authentication ceremony operations.

This is broadly modelled after the functions in the [@simplewebauthn/server](https://simplewebauthn.dev/docs/packages/server) npm package.

yawaramin committed:
Add `Simple` module

Add a module that introduces simplified registration & authentication
ceremony operations.

This is broadly modelled after the functions in the
[@simplewebauthn/server](https://simplewebauthn.dev/docs/packages/server)
package.

yawaramin committed:
Remove redundant parameters

Remove parameters that actually don't make sense for the current state
of this library.

yawaramin committed:
Adjust docstring wording

reynir commented on src/webauthn.ml:
```diff
@@ -480,3 +481,137 @@ let transports_of_cert c =
+    (if expected_challenge |> challenge_equal challenge |> not
+    then invalid_arg "verify_registration_response: challenge mismatch");
+
+    let created_at = Unix.time () in
```
I'd like this to be an argument to the function so we don't depend on Unix (and thus can use it on Mirage) :-)

yawaramin committed:
Docstring formatting

reynir commented on src/webauthn.mli:
```diff
@@ -184,3 +189,215 @@ val decode_transport : string -> (transport list, [> `Msg of string ]) result
+    aaguid : string;
+
+    counter : Int32.t;
+    (** Increment this after each successful authentication ceremony. *)
```
Oh no, isn't this the counter received from the authenticator? In that case it's the authenticator that increments it likely across all sites. Or always return zero. See this part of the demo: https://github.com/robur-coop/webauthn/blob/22c720b2a5c018852fb718b75487da523dc2a282/bin/webauthn_demo.ml#L33-L59

reynir commented on src/webauthn.mli:
```diff
@@ -184,3 +189,215 @@ val decode_transport : string -> (transport list, [> `Msg of string ]) result
+    (** Foreign key referencing the users table. *)
+
+    pub_key : pub_key;
+    aaguid : string;
```
I think this needs to be an option - clients don't always attest - and the web server can suggest whether it wants an attestation https://developer.mozilla.org/en-US/docs/Web/API/PublicKeyCredentialCreationOptions#attestation

yawaramin replied:
This is coming from `registration.attested_credential_data.aaguid` so should that also be a `string option`?

reynir replied:
The whole `registration.attested_credential_data` is an `_ option`

reynir replied:
Yea I see. But the authenticator can decide not to do an attestation.

Testing locally the `aaguid` on Firefox ESR 128 seems to be always all-zeroes with either a yubikey or a feitian multipass. I don't know if this is because I use origin `http://localhost:5000`?! I'm a bit confused by this. It's also unclear if this is due to me using an older version of Firefox (thus older WebAuthn spec)

reynir replied:
It seems there are substantive changes since "level 2" that changes the semantics of aaguid. I probably need to read up on that again... https://w3c.github.io/webauthn/#changes-l3-substantive

reynir replied:
Also relevant: https://w3c.github.io/webauthn/#ref-for-authdata-attestedcredentialdata-aaguid%E2%91%A0

yawaramin replied:
I have ported the demo app on top of the `Simple` module if you want to try it out: https://github.com/yawaramin/webauthn/tree/demo-ui-ux/bin

Can run with `dune exec ./bin/app.exe` from the project root.

yawaramin replied:
It looks like Apple and Google are not providing attestations any more: https://news.ycombinator.com/item?id=42522490

reynir commented on src/webauthn.mli:
```diff
@@ -184,3 +189,215 @@ val decode_transport : string -> (transport list, [> `Msg of string ]) result
+    (** Use this as the lookup key when storing passkeys. *)
+
+    user_id : string;
+    (** Foreign key referencing the users table. *)
```
Hmm, is it?

reynir replied:
It should be what you pass as`~user_id` to `generate_registration_options`, and the default is a random byte sequence EDIT: in this implementation, I mean.

reynir commented on src/webauthn.ml:
```diff
@@ -480,3 +481,137 @@ let transports_of_cert c =
+      |> or_invalid
+    in
+    (if expected_challenge |> challenge_equal challenge |> not
+    then invalid_arg "verify_registration_response: challenge mismatch");
```
We otherwise use result. I would prefer if we keep using result. I'm okay with a `verify_registration_response_exn` in addition, though.

reynir commented on src/webauthn.ml:
```diff
@@ -480,3 +481,137 @@ let transports_of_cert c =
+    }
+
+  let generate_authentication_options
+    ?(allow_credentials=[])
```
I think having `allow_credentials` an optional argument that defaults to the empty list is not so nice. You are supposed to put the credentials the user can use here. If you put none the client can't authenticate.

yawaramin replied:
Yojson will just not emit this field in the JSON object if it's left as the default empty list. So actually any credential will be allowed.

reynir replied:
I think this would be nice to add a comment about because I will almost definitely forget this behavior next time I look at the code :)

yawaramin committed:
Adjust docstring wording

yawaramin committed:
Extract base-64 encoding function

Since we are repeatedly using URL-safe base-64 encoding.

reynir commented on src/webauthn.mli:
```diff
@@ -29,13 +31,16 @@
 
-(** [create origin] is a webauthn state, or an error if [origin] does not
+(** [create ?name origin] is a webauthn state for a relying party [name] (default
+    ["localhost"]) and hostname [origin], or an error if the [origin] does not
```
So I still need to figure out what the situation is with `name`, but a better default might be the rpid derived from `origin`.
https://w3c.github.io/webauthn/#dom-publickeycredentialentity-name

reynir commented on src/webauthn.mli:
```diff
@@ -184,3 +189,215 @@ val decode_transport : string -> (transport list, [> `Msg of string ]) result
+      const credential = await navigator.credentials.get({ publicKey: options });
+      ]} *)
+
+  val verify_authentication_response :
+    expected_challenge:challenge ->
+    pub_key:pub_key ->
+    string ->
+    t ->
+    authentication
+  (** [verify_authentication_response ~expected_challenge ~pub_key response
+      webauthn] is an [authentication] object constructed after verifying the
+      authentication response.
+
+      The [response] can be obtained with something like this:
+
+      {@javascript[
+      const credential = await navigator.credentials.get({ publicKey: options });
+      const response = JSON.stringify(credential.toJSON().response);
+      // Upload credential.id and response to server
+      ]}
+
+      The [pub_key] can be obtained by looking up the stored {!passkey}
+      corresponding to [credential.id] obtained from the client, and getting its
+      public key.
+
+      @raise Invalid_argument if authentication verification fails. *)
```
I think it's a bit unclear what `response` is. The javascript comment says to upload `credential.id` and `response` to the server. But it looks like the `response` to this function is the javascript `response` value. In the demo we send the `credential.id` in the URL path (I think), but reading this you might think you put the `credential.id` in the response to the server somehow. I think it could be nice if this was used in the demo.

I think it's also worth mentioning that this doesn't check the counter at all. This is a valid choice, but not necessarily what the user of this library wants or expects.

reynir reviewed (commented):
I like the `generate_registration_options` and `generate_authentication_options` functions. If I remember correctly these JSON "types" were not standardized yet, and they would be nice to have.

As for the `passkey` type and `verify_registration_response` and `verify_authentication_response` I'm somewhat undecided.

Perhaps we can JSON related functions and types into the top level module? I would be happy to merge that. Otherwise I need more time to review this and I need to focus on something else in the near future.

hannesm commented on src/dune:
```diff
@@ -3,4 +3,4 @@
  (preprocess
   (pps ppx_deriving_yojson))
- (libraries mirage-crypto-rng yojson mirage-crypto-ec x509 base64 cbor))
+ (libraries mirage-crypto-rng yojson mirage-crypto-ec x509 base64 cbor unix))
```
please let us keep the core library free of unix, so we can use it in MirageOS.

hannesm commented on src/webauthn.ml:
```diff
@@ -480,3 +483,135 @@ let transports_of_cert c =
+    | Ok s -> s
+    | Error (`Msg m) -> failwith m
+
+  let challenge_to_yojson c = `String (c |> challenge_to_string |> b64_urlenc)
```
```suggestion
  let challenge_to_yojson c = `String (challenge_to_string c |> b64_urlenc)
```

It looks like you like the pipeline operator `|>` very much. I don't like it that much, and would use it where it makes the code more readable. IMHO this is not the case here.

hannesm commented on src/webauthn.ml:
```diff
@@ -480,3 +483,135 @@ let transports_of_cert c =
+  type pub_key = Mirage_crypto_ec.P256.Dsa.pub
+
+  let pub_key_to_yojson pk = `String (
+    pk
+    |> Mirage_crypto_ec.P256.Dsa.pub_to_octets
```
```suggestion
    Mirage_crypto_ec.P256.Dsa.pub_to_octets pk
```

hannesm commented on src/webauthn.ml:
```diff
@@ -480,3 +483,135 @@ let transports_of_cert c =
+
+  let pub_key_of_yojson = function
+    | `String s ->
+      (match Mirage_crypto_ec.P256.Dsa.pub_of_octets (b64_urldec s) with
```
and here, suddenly, you don't use the pipeline operator? I'm in favor to stick to one style.

hannesm commented on src/webauthn.ml:
```diff
@@ -480,3 +483,135 @@ let transports_of_cert c =
+    | `String s ->
+      (match Mirage_crypto_ec.P256.Dsa.pub_of_octets (b64_urldec s) with
+      | Ok pk -> Ok pk
+      | Error _ -> Error "invalid public key")
```
the error message may have been useful, but you throw it away....

hannesm commented on src/webauthn.ml:
```diff
@@ -480,3 +483,135 @@ let transports_of_cert c =
+      |> register webauthn
+      |> or_invalid
+    in
+    (if expected_challenge |> challenge_equal challenge |> not
```
```suggestion
    (if  not (challenge_equal challenge expected_challenge)
```

I'm sorry, but with the pipeline operator for me it becomes very hard to digest.

hannesm commented on src/webauthn.ml:
```diff
@@ -480,3 +483,135 @@ let transports_of_cert c =
+    | Error e -> Fmt.kstr invalid_arg "error: %a" pp_error e
+
+  let verify_registration_response ~expected_challenge ~user_id response webauthn =
+    let challenge, registration = response
+      |> register_response_of_string
```
```suggestion
    let challenge, registration =
      register_response_of_string response
```

hannesm commented on src/webauthn.ml:
```diff
@@ -480,3 +483,135 @@ let transports_of_cert c =
+  }
+
+  let verify_authentication_response ~expected_challenge ~pub_key response webauthn =
+    let challenge, authentication = response
+      |> authenticate_response_of_string
```
```suggestion
    let challenge, authentication =
      authenticate_response_of_string response
```

hannesm commented on src/webauthn.ml:
```diff
@@ -480,3 +483,135 @@ let transports_of_cert c =
+      |> authenticate webauthn pub_key
+      |> or_invalid
+    in
+    (if expected_challenge |> challenge_equal challenge |> not
```
```suggestion
    (if not (challenge_equal challenge expected_challenge)
```

hannesm reviewed (commented):
I mainly have stylistic comments

yawaramin commented:
Guys, thanks for the reviews. I will fix all these issues later today.

yawaramin committed:
Address code review feedback

- Remove `unix` dependency to allow Mirage
- Return results instead of raising exceptions
- Do signature counter verification automatically
- Use let-operator instead of pipe
- Use origin as default name
- Preserve Mirage crypto public key deserialize error message
- Make `user_id` parameter mandatory so that we always preserve
  correlation with user database
- Improve docstring example

yawaramin committed:
Adjust docstrings

yawaramin committed:
Docstring markup

reynir commented on src/webauthn.ml:
```diff
@@ -280,7 +289,7 @@ let create origin =
         | _ -> Error ("invalid origin host and port " ^ host_port)
       with
-      | Ok host -> Ok { origin ; rpid = host }
+      | Ok host -> Ok { name = Option.value ~default:origin name ; origin ; rpid = host }
```
It should be the host (without http(s)://)
```suggestion
      | Ok host -> Ok { name = Option.value ~default:host name ; origin ; rpid = host }
```

reynir commented on src/webauthn.ml:
```diff
@@ -23,6 +23,8 @@ type error = [
   | `Rpid_hash_mismatch of string * string
   | `Missing_credential_data
   | `Signature_verification of string
+  | `Challenge_mismatch of string * string
+  | `Sign_count_mismatch of Int32.t * Int32.t
```
Maybe we can add a ``type verification_error = [ error | `Challenge_mismatch | Sign_count_mismatch ]`` so we don't have those two constructors in the result of the existing functions.

yawaramin replied:
The benefit of having a single type is that we only need a single `pp_error` to be able to print any error for debugging purposes. If we add more then we need to hunt for the correct `pp` to be able to print them. Could get annoying.

reynir commented on src/webauthn.ml:
```diff
@@ -480,3 +489,126 @@ let transports_of_cert c =
+      | Ok octets ->
+        (match Mirage_crypto_ec.P256.Dsa.pub_of_octets octets with
+        | Ok _ as pub_key -> pub_key
+        | Error e -> Fmt.error "%a" Mirage_crypto_ec.pp_error e)
```
Minor: let's bind the `Ok _` thing as `r`. If we bind it as `pub_key` I tend to think it's `Ok pub_key` (that is, `pub_key` is the actual public key).
```suggestion
        (match Mirage_crypto_ec.P256.Dsa.pub_of_octets octets with
        | Ok _ as r -> r
        | Error e -> Fmt.error "%a" Mirage_crypto_ec.pp_error e)
```
It could also be written using `Result.map_error`

```suggestion
        Mirage_crypto_ec.P256.Dsa.pub_of_octets octets
        |> Result.map_error (Fmt.to_to_string Mirage_crypto_ec.pp_error)
```

yawaramin committed:
Minor tweaks

- Default name should be the hostname without protocol
- Adjust pattern match
- Adjust docstring

yawaramin committed:
Allow filtering credentials by transport

Also hint at storing challenge in code examples.

yawaramin committed:
Add display name to passkey record

Heyji2 commented:
> Add a module that introduces simplified registration & authentication ceremony operations.
> 
> This is broadly modelled after the functions in the [@simplewebauthn/server](https://simplewebauthn.dev/docs/packages/server) npm package.

I think this new interface is great, much easier to understand. However the whole purpose of the @simplewebauthn library is the client/server pair, which simplifies the transfer of array buffers from the client to the server. This is explained in the [Philosophy](https://simplewebauthn.dev/docs/simplewebauthn/philosophy) pargraph of the documentation. 

So Ideally, one would need the client side as well, so as to simplify developer's use of this library. As such, this library is not compatible with @simplewebauthn/browser.

yawaramin commented:
Hi @Heyji2, actually the client side is not needed any more, because the Web platform APIs support JSON/ArrayBuffer encoding/decoding now. See the docstrings I added for some examples.

Or alternatively, this backend API should actually 'just work' with the `@simplewebauthn/browser` package as well. Here's an example that uses the Python `webauthn` library with the `@simplewebauthn/browser` package directly: https://github.com/AnswerDotAI/fasthtml-example/blob/bd24075cfba81ed949a43edfa5db6c67eb84a6c5/passkey/main.py#L83

Heyji2 commented:
Indeed, I am using it in a personal project, and I confirm, it works with `@simplewebauthn/browser`

Heyji2 commented:
@yawaramin, if I may, your `verify_registration_response` returns a `passkey result` type. Wouldn't it be wise to follow the [norm](https://www.w3.org/TR/webauthn-3/#sctn-registering-a-new-credential)'s identifiers and structures, so that is returns instead a 
`credentialRecord` structure (step 27), with almost the same fields (but with different names) that could then be stored on the `user-account` on the server as-is with Yojson serializers as explained on step 29 ?

yawaramin commented:
@Heyji2 that might be doable, but how to get these field values?

- `uvInitialized`
- `backupEligible`
- `backupState`

Heyji2 commented:
These fields are to be taken from the `flags` field of the `authenticator data` of `AuthenticatorAttestationResponse`. This is described in steps 15 to 19. If they are not yet made available by the underlying layer, I guess you can leave them empty for the time being.
